ruma-api-macros/src/util.rs

```diff
@@ -2,10 +2,62 @@
 
 use proc_macro2::{Span, TokenStream};
 use quote::quote;
-use syn::Ident;
+use std::collections::BTreeSet;
+use syn::{
+    AngleBracketedGenericArguments, GenericArgument, Ident, Lifetime,
+    ParenthesizedGenericArguments, PathArguments, Type, TypePath, TypeReference,
+};
 
 use crate::api::{metadata::Metadata, request::Request};
 
+pub fn copy_lifetime_ident(lifetimes: &mut BTreeSet<Lifetime>, ty: &Type) {
+    match ty {
+        Type::Path(TypePath { path, .. }) => {
+            for seg in &path.segments {
+                match &seg.arguments {
+                    PathArguments::AngleBracketed(AngleBracketedGenericArguments {
+                        args, ..
+                    }) => {
+                        for gen in args {
+                            if let GenericArgument::Type(ty) = gen {
+                                copy_lifetime_ident(lifetimes, &ty)
+                            } else if let GenericArgument::Lifetime(lt) = gen {
+                                lifetimes.insert(lt.clone());
+                            }
+                        }
+                    }
+                    PathArguments::Parenthesized(ParenthesizedGenericArguments {
+                        inputs, ..
+                    }) => {
+                        for ty in inputs {
+                            copy_lifetime_ident(lifetimes, ty)
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+        Type::Reference(TypeReference { elem, lifetime, .. }) => {
+            copy_lifetime_ident(lifetimes, &*elem);
+            if let Some(lt) = lifetime {
+                lifetimes.insert(lt.clone());
+            }
+        }
+        _ => {}
```

Comment:
Would be nice to handle everything that can (transitively) contain a lifetime here (so tuples and such), but that can be done in a followup PR too.

Reply:
Are there any others that could reasonably be found? [syn::Type](https://docs.rs/syn/0.15.44/syn/enum.Type.html)

Reply:
I think all of them except `Never`, `Infer`, `Macro` and `Verbatim` should ideally be recursed into. Although many are never realistically going to show up inside our code, e.g. `TypeImplTrait` or `TypeBareFn`. Slices, tuples, `TypeParen` and `TypeGroup` are most likely to be encountered.